specification/dartLangSpec.tex

```diff
@@ -13654,16 +13657,30 @@ \subsubsection{Ordinary Invocation}
 
 \LMHash{}%
 It is a compile-time error to invoke any of the methods of class \code{Object}
-on a constant type literal that is immediately followed by the token `.'\,.
+on a type literal that is immediately followed by the token `.' (a period).
```

Comment:
Should we mention `?.` as well.
`C?.toString()` is "treated like" `C.toString()`, but it is not "immediately followed by the token `.'".
I could argue that it's not a compile time error, and then it's treated like `C.toString()` would be as a non-compile-time error. (I can argue differently too, but that's still bad).

Or should we just remove this, and just rely on the rule that if `C` is a type literal then `C.m` is a compile time error if `C` does not declare a static member named `m`.
(This text comes from back when static lookups were lookups on objects with the static members as instance members. It never worked.)

Basically (which is what I think this CL intends, just summarizing):

If `T` is a type literal or denotes an extension, then the conditional property extraction `T?.m` is equivalent to `T.m` (and the conditional method invocation `T?.m()` is equivalent to `T.m()`).
If `T` is a type literal denoting a class or mixin `C`, or denoting an extension `E`, then `T.m` and `T.m()` are compile-time errors if `C`, respectively `E`, does not declare a static member named `m`.

(Come constructor tear-offs, make it "a static member or constructor with base name `m`", and only allow static accesses on non-instantiated type literals).

Then make the ordinary method invocation/property extraction `e.m` split into either when `e` is a type literal, or when it's not, and have separate behaviors for those.


Then it's a compile-time error to write `C.toString()` not (directly) because `Object` declares `toString`, but because `C` does not declare a static with that name (which it can't because it cannot declare a static with the same name as an inherited instance member, so only indirectly because it's declared by `Object`).

The result should be the same, but without the "one token lookahead" which doesn't really scale.

Reply:
If we just remove the entire paragraph, then I don't think there is any problem at all.
Or make it descriptive instead of nominative.

The semantics of `C.m` and `C.m()` where `C` is a type literal are specified such that we don't even *try* to look up an instance member, only a static member (or constructor for invocations), and we also won't go to the extension method phase which comes after failing to find an instance member.


We'll definitely need to revisit a lot of things :)